src/coreclr/vm/jitinterface.cpp

```diff
@@ -1514,7 +1514,22 @@ void CEEInfo::getFieldInfo (CORINFO_RESOLVED_TOKEN * pResolvedToken,
             CORINFO_FIELD_ACCESSOR intrinsicAccessor;
 
             if (pField->GetFieldType() == ELEMENT_TYPE_VALUETYPE)
-                fieldFlags |= CORINFO_FLG_FIELD_STATIC_IN_HEAP;
+            {
+                bool frozenBoxedStatic = false;
+                if (pFieldMT->ContainsPointersOrCollectible())
+                {
+                    Object** handle = (Object**)pField->GetStaticAddressHandle(pField->GetBase());
+                    if (*handle == nullptr && GCHeapUtilities::GetGCHeap()->IsInFrozenSegment(*handle))
+                    {
+                        frozenBoxedStatic = true;
+                    }
+                }
+
+                if (!frozenBoxedStatic)
+                {
+                    fieldFlags |= CORINFO_FLG_FIELD_STATIC_IN_HEAP;
+                }
+            }
 
```

Comment:
I do not think that this will work for shared generic code. Shared generic code needs to work for both collectible and non-collectible generic instantiations, so you have to assume that it is always `CORINFO_FLG_FIELD_STATIC_IN_HEAP`